decidim/decidim
# Fix memory leak with user answers serializer (at survey export)

#### :tophat: What? Why?
The form user answer serializer has a memory leak when used e.g. at the surveys component export functionality.

You can see this memory leak when the following conditions apply:
- The questionnaire has a very long description
- The survey/questionnaire has a large amount of answers

This memory leak happens because the `questionnaire` is fetched for each "answer set" in the list of user answers which looks like this:

```
[
  # Answers for user A
  [#<Decidim::Forms::Answer:0x0000000000000001 id: 1, ...>, #<Decidim::Forms::Answer:0x0000000000000002 id: 2, ...>],
  # Answers for user B
  [#<Decidim::Forms::Answer:0x0000000000000003 id: 3, ...>, #<Decidim::Forms::Answer:0x0000000000000004 id: 4, ...>],
  # ...
]
```

When you iterate over this array and for each "answer set" in the top-level array call `answer_set.first.questionnaire`, the questionnaire is loaded into memory along with its description causing the memory leak.

#### Testing
- Create a survey with a very long description (make it reeeeeeeeeaallly long)
- Add 2000 answers to that survey
- Follow your programs using `top` in another terminal
- Start the export in Decidim
- See the memory usage of the exporting process growing all of the time until it runs out of memory and stops

Do this test with the current version without this fix and you should expect to see the memory leak, i.e. the report will not be generated correctly. If you apply this fix, the error problem should disappear.
